[SPARK-47365][PYTHON] Add toArrow() DataFrame method to PySpark

ianmcook opened the pull request:
### What changes were proposed in this pull request?
- Add a PySpark DataFrame method `toArrow()` which returns the contents of the DataFrame as a [PyArrow Table](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html), for both local Spark and Spark Connect.
- Add a new entry to the **Apache Arrow in PySpark** user guide page describing usage of the `toArrow()` method.
- Add  a new option to the method `_collect_as_arrow()` to provide more useful output when there are zero records returned. (This keeps the implementation of `toArrow()` simpler.)

### Why are the changes needed?
In the Apache Arrow community, we hear from a lot of users who want to return the contents of a PySpark DataFrame as a PyArrow Table. Currently the only documented way to do this is to return the contents as a pandas DataFrame, then use PyArrow (`pa`) to convert that to a PyArrow Table.
```py
pa.Table.from_pandas(df.toPandas())
```
But going through pandas adds significant overhead which is easily avoided since internally `toPandas()` already converts the contents of Spark DataFrame to Arrow format as an intermediate step when `spark.sql.execution.arrow.pyspark.enabled` is `true`.

Currently it is also possible to use the experimental `_collect_as_arrow()` method to return the contents of a PySpark DataFrame as a list of PyArrow RecordBatches. This PR adds a new non-experimental method `toArrow()` which returns the more user-friendly PyArrow Table object.

This PR also adds a new argument `empty_list_if_zero_records` to the experimental method `_collect_as_arrow()` to control what the method returns in the case when the result data has zero rows. If set to `True` (the default), the existing behavior is preserved, and the method returns an empty Python list. If set to `False`, the method returns returns a length-one list containing an empty Arrow RecordBatch which includes the schema. This is used by `toArrow()` which requires the schema even if the data has zero rows.

For Spark Connect, there is already a `SparkSession.client.to_table()` method that returns a PyArrow table. This PR uses that to expose `toArrow()` for Spark Connect.

### Does this PR introduce _any_ user-facing change?

- It adds a DataFrame method `toArrow()` to the PySpark SQL DataFrame API.
- It adds a new argument `empty_list_if_zero_records` to the experimental DataFrame method `_collect_as_arrow()` with a default value which preserves the method's existing behavior.
- It exposes `toArrow()` for Spark Connect, via the existing `SparkSession.client.to_table()` method.
- It does not introduce any other user-facing changes.

### How was this patch tested?
This adds a new test and a new helper function for the test in `pyspark/sql/tests/test_arrow.py`.

### Was this patch authored or co-authored using generative AI tooling?
No

ianmcook committed:
Add _toArrow()

ianmcook committed:
Remove unnecessary try

ianmcook committed:
Add empty_list_if_zero_records option to _collect_as_arrow()

ianmcook committed:
Add missing imports

ianmcook committed:
Format Python

HyukjinKwon commented on python/pyspark/sql/pandas/conversion.py:
```diff
@@ -241,21 +241,81 @@ def toPandas(self) -> "PandasDataFrameLike":
         else:
             return pdf
 
-    def _collect_as_arrow(self, split_batches: bool = False) -> List["pa.RecordBatch"]:
```
This isn't an API; I wouldn't fix it for external purposes.

ianmcook replied:
Ok, understood. I will make some changes here to reflect that.

P.S. The git diff here is confusing and makes it look like I am proposing a rename and full reimplementation of `_collect_as_arrow` which I am not :)

ianmcook committed:
Changes based on review feedback

ianmcook committed:
Format Python

ianmcook committed:
Add test

ianmcook committed:
Fix timezone error in test

ianmcook committed:
Resolve rebase conflicts

ianmcook committed:
Bugfix

ianmcook commented:
cc @xinrong-meng @itholic @dongjoon-hyun

ianmcook commented on python/pyspark/sql/connect/dataframe.py:
```diff
@@ -1775,6 +1775,12 @@ def _to_table(self) -> Tuple["pa.Table", Optional[StructType]]:
+        table = self._to_table()[0]
+        return table
+
+    _toArrow.__doc__ = PySparkDataFrame._toArrow.__doc__
```
@HyukjinKwon is this line no longer needed after https://github.com/apache/spark/pull/46129?

ianmcook committed:
Remove unneeded __doc__

grundprinzip commented on python/pyspark/sql/connect/dataframe.py:
```diff
@@ -1775,6 +1775,10 @@ def _to_table(self) -> Tuple["pa.Table", Optional[StructType]]:
         return (table, schema)
 
+    def _toArrow(self) -> "pa.Table":
+        table = self._to_table()[0]
```
nit: Using this notion, makes the code a bit more readable and makes sure it fails if _to_table ever changes the tuple size, and thus behavior.

```suggestion
        table, _ = self._to_table()
```

grundprinzip commented on python/pyspark/sql/connect/dataframe.py:
```diff
@@ -1775,6 +1775,10 @@ def _to_table(self) -> Tuple["pa.Table", Optional[StructType]]:
         assert table is not None
         return (table, schema)
 
+    def _toArrow(self) -> "pa.Table":
```
@HyukjinKwon I think it would be fine to make this a regular public method instead of the `_toArrow` maybe change it to `toArrowTable` to give a better indication of the return type.

HyukjinKwon replied:
Yeah, the current state looks a bit weird. We should either explicitly add an developer API, or just don't. My only question is that we can work around via `SparkSession.client.to_table()`, and we're going forward to Spark Connect default. Considering this is a developer API, I wonder if we should add this into DataFrame at this moment.

ianmcook replied:
I updated the PR to make the method public and rename it `toArrowTable`.

I will also push a commit here that adds some relevant content to the [Apache Arrow in PySpark](https://spark.apache.org/docs/latest/api/python/user_guide/sql/arrow_pandas.html) page.

ianmcook replied:
I expect that this will be used by some end users, not only by developers. In recent years, PyArrow has become a very widely used library (>100M downloads/month on PyPI, ~50th most popular library) and we see a lot of non-developers using it.

ianmcook committed:
Incorporate feedback from review

Co-authored-by: Martin Grund <grundprinzip@gmail.com>

ianmcook committed:
Make non-experimental and rename toArrowTable

ianmcook committed:
Add entry to Apache Arrow in PySpark page

ianmcook committed:
Add missing import in example

ianmcook committed:
Ignore unused pyarrow import

ianmcook committed:
Add additional space before # noqa

ianmcook committed:
Try fix mypy check error

xinrong-meng commented on python/pyspark/sql/pandas/conversion.py:
```diff
@@ -225,15 +225,68 @@ def toPandas(self) -> "PandasDataFrameLike":
+
+        import pyarrow as pa
+
+        self_destruct = jconf.arrowPySparkSelfDestructEnabled()
```
Shall we document under "Setting Arrow self_destruct for memory savings" of `python/docs/source/user_guide/sql/arrow_pandas.rst` that Conversion to Arrow Table will also be influenced by the config?
Following "which can save memory when creating a Pandas DataFrame via toPandas..." for example.

ianmcook replied:
Added in fd76fa3. Thank you.

ianmcook committed:
Document use of self_destruct with toArrowTable

HyukjinKwon commented:
Okay I'm fine with this

ianmcook commented:
Thanks @HyukjinKwon. Anything else you need before merging this?

HyukjinKwon commented on python/pyspark/sql/connect/dataframe.py:
```diff
@@ -1775,6 +1775,10 @@ def _to_table(self) -> Tuple["pa.Table", Optional[StructType]]:
         assert table is not None
         return (table, schema)
 
+    def toArrowTable(self) -> "pa.Table":
```
Can we list this to `python/docs/source/reference/pyspark.sql/dataframe.rst`?

HyukjinKwon replied:
Let's also add the abstract method at the parent dataframe `pyspark.sql.dataframe`

ianmcook replied:
> Let's also add the abstract method at the parent dataframe `pyspark.sql.datafram`

I don't see where to do this. Can you clarify please? Thank you

HyukjinKwon replied:
Here https://github.com/apache/spark/blob/master/python/pyspark/sql/dataframe.py :-)

HyukjinKwon replied:
I think you should probably rebase it against the master branch. That refactoring happened lately (by me .. )

ianmcook replied:
Thanks.  I think I added what's needed there. It's a placeholder definition, like what you have there for `toPandas`.  See my other comment below.

HyukjinKwon reviewed (approved):
LGTM otherwise

ianmcook committed:
Add entry to dataframe.rst

ianmcook committed:
Update for consistency after #46129

ianmcook commented:
Thanks. I rebased. I also took a closer look at your changes in #46129 and made a few changes for consistency with the new structure you introduced there:
- I added a definition for `toArrowTable` in `python/pyspark/sql/classic/dataframe.py`.
- I added a placeholder definition for `toArrowTable` in `python/pyspark/sql/dataframe.py`
- I moved the docstring to there from `python/pyspark/sql/pandas/conversion.py`.

I also added `.. versionadded:: 4.0.0` in the docstring.

Please let me know if that all looks OK.

ianmcook commented on python/pyspark/sql/dataframe.py:
```diff
@@ -6213,6 +6214,31 @@ def mapInArrow(
         """
         ...
 
```
Do I need ` @dispatch_df_method` here?

HyukjinKwon replied:
yes please

HyukjinKwon commented on python/docs/source/reference/pyspark.sql/dataframe.rst:
```diff
@@ -109,6 +109,7 @@ DataFrame
     DataFrame.tail
     DataFrame.take
     DataFrame.to
+    DataFrame.toArrowTable
```
One last thing. Should we name it `toArrow` just to match with others like `toPandas`, `toJSON`, etc.? I don't have a strong preference though.

HyukjinKwon replied:
Yeah, let's probably name it `toArrow` just to match with others, e.g., https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.to_pandas too

ianmcook replied:
Either name is fine by me, but @grundprinzip  commented that he prefers `toArrowTable ` https://github.com/apache/spark/pull/45481#discussion_r1592074090

HyukjinKwon replied:
Let's go with toArrow. I think this kind of naming is more common, e.g., https://datatable.readthedocs.io/en/latest/api/frame/to_arrow.html https://docs.pola.rs/py-polars/html/reference/dataframe/api/polars.DataFrame.to_arrow.html

ianmcook replied:
Sounds good. Done.

HyukjinKwon commented on python/pyspark/sql/dataframe.py:
```diff
@@ -6213,6 +6214,31 @@ def mapInArrow(
+        Notes
+        -----
+        This method should only be used if the resulting PyArrow ``pyarrow.Table`` is
+        expected to be small, as all the data is loaded into the driver's memory.
```
```suggestion
        expected to be small, as all the data is loaded into the driver's memory.

        This API is a developer API.
```

ianmcook committed:
Add @dispatch_df_method

ianmcook committed:
Incorporate feedback from review

Co-authored-by: Hyukjin Kwon <gurwls223@gmail.com>

ianmcook committed:
Rename toArrowTable -> toArrow

HyukjinKwon committed:
Update dataframe.py

HyukjinKwon committed:
Update dataframe.py

HyukjinKwon committed:
Update examples/src/main/python/sql/arrow.py
